servers/src/main/java/tachyon/worker/block/evictor/LRUEvictor.java

```diff
@@ -83,12 +126,64 @@ public EvictionPlan freeSpace(long availableBytes, BlockStoreLocation location)
     }
 
     // enough free space
-    if (dirCandidates.maxAvailableBytes() >= toEvictBytes) {
-      toEvict = dirCandidates.toEvict();
-      for (Long blockId : toEvict) {
-        mLRUCache.remove(blockId);
+    if (dirCandidates.candidateSize() >= bytesToBeAvailable) {
+      // candidate blockIds for eviction from current tier, sorted from less recently used to more
+      // recently used
+      toEvict = dirCandidates.candidateBlocks();
+      // reverse list so that the more recently used blocks are moved to next tier first
+      Collections.reverse(toEvict);
+      // TODO: maybe we should abstract the strategy of moving blocks to lower tier
+      // move as many blocks to next tier as possible
+      List<StorageTier> tiersBelow =
+          mMeta.getTiersBelow(dirCandidates.candidateDir().getParentTier().getTierAlias());
+      for (StorageTier tier : tiersBelow) {
+        for (StorageDir dir : tier.getStorageDirs()) {
+          BlockStoreLocation dest = dir.toBlockStoreLocation();
+          Iterator<Long> blocks = toEvict.iterator();
+          long remainBytes = dir.getAvailableBytes();
+          while (blocks.hasNext() && remainBytes >= 0) {
+            long blockId = blocks.next();
+            long blockSize = mMeta.getBlockMeta(blockId).getBlockSize();
+            if (blockSize <= remainBytes) {
+              // the block can be moved to the dir
+              toMove.add(new Pair<Long, BlockStoreLocation>(blockId, dest));
+              blocks.remove();
+              remainBytes -= blockSize;
+            }
+          }
+        }
+      }
+
+      // assure all blocks are in the store, if not, remove from plan and lru cache
+      long toFree = 0L;
+      Iterator<Pair<Long, BlockStoreLocation>> moveIt = toMove.iterator();
+      while (moveIt.hasNext()) {
+        long id = moveIt.next().getFirst();
+        if (!mMeta.hasBlockMeta(id)) {
+          moveIt.remove();
+          mLRUCache.remove(id);
+        } else {
+          toFree += mMeta.getBlockMeta(id).getBlockSize();
+        }
+      }
+      Iterator<Long> evictIt = toEvict.iterator();
+      while (evictIt.hasNext()) {
+        long id = evictIt.next();
+        if (!mMeta.hasBlockMeta(id)) {
+          evictIt.remove();
+          mLRUCache.remove(id);
+        } else {
+          toFree += mMeta.getBlockMeta(id).getBlockSize();
+        }
+      }
+
+      // reassure the plan is feasible
+      if (mMeta.getAvailableBytes(location) + toFree >= bytesToBeAvailable) {
+        for (Long blockId : toEvict) {
+          mLRUCache.remove(blockId);
```

Comment:
I don't think you need to call remove here, since they onRemove callback will be triggered when the block is actually evicted.